### PR TITLE
docs(*): add documentation and reST files for readthedocs

### PR DESCRIPTION
**Pull request by zigzagcai**
## Motivation

Add readthedocs for InternLM

## Modification

Preview of readthedocs private repo can be accessed [here](https://czinternlm.readthedocs.io/en/latest/).

1. we added reST files and corresponding images in `docs/**`
2. minor modification with comments in `internlm/**`, making it compatible with Sphinx autoclass and autofunction
3. reuse `install.md` and `usage.md` and set `myst-parser` extension to render

## BC-breaking (Optional)

Not involved.

## Use cases (Optional)

Not involved.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: doc/code-docs/source/parallel.rst [hunk 1]
```diff
@@ -3,21 +3,101 @@ Parallel Training
 
 .. 整体说一下并行配置使用方式，接下来再分模块详细说明
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:4]: 这个注释可以改成英文

File: doc/code-docs/source/parallel.rst [hunk 2]
```diff
@@ -3,21 +3,101 @@ Parallel Training
...
 
+InternLM supports tensor parallel, pipeline parallel, sequence parallel, data parallel, and ZeRO1.5 to parallelize the training pipeline. 
+When initializing the distributed environment, we need to specify tensor parallel size, pipeline parallel size, data parallel size, 
+and ZeRO1.5 strategy. 
+
+The parallel setting of InternLM is fully config-driven, and you can change the parallelism by modifying 
+`config file <https://github.com/InternLM/InternLM/blob/main/configs/7B_sft.py>`_.
+
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:12]: 这里可以把parallel那块配置贴出来

File: doc/code-docs/source/parallel.rst [hunk 3]
```diff
@@ -3,21 +3,101 @@ Parallel Training
...
 -----------------
 
+InternLM uses `1F1B <https://arxiv.org/pdf/2104.04473.pdf>`_ (one forward pass followed by one backward pass) for pipeline parallel. For 1F1B strategy, there are two implementations: 
+(1) non-interleaved scheduler, which is memory-efficient (2) interleaved scheduler, which is both memory-efficient and time-efficient.
+
+.. figure:: ../../imgs/pipeline_schedule.png
+  :scale: 45%
+  :class: with-border
+
+  Non-interleaved and interleaved scheduler for 1F1B pipeline parallelism, adopted from Megatron-LM original paper
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:40]: 这里Megatron-LM可以链接一下论文地址，上面flash-attn也是

File: doc/code-docs/source/parallel.rst [hunk 4]
```diff
@@ -3,21 +3,101 @@ Parallel Training
...
+.. figure:: ../../imgs/sequence_parallel.png
+  :scale: 50%
+  :class: with-border
+
+  Sequence parallel, adopted from flash-attention original paper
 
 Data Parallel
 -----------------
 
+InternLM supports both vanilla data parallel and ZeRO1.5 (an optimized implementation of ZeRO).
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:76]: vanilla是指什么

File: doc/code-docs/source/parallel.rst [hunk 5]
```diff
@@ -3,21 +3,101 @@ Parallel Training
...
+2. If ``parallel.zero1 == 1``, zero is not used, and all dp groups retain the full amount of model parameters.
+3. If ``parallel.zero1 > 1`` and ``parallel.zero1 <= dp world size``, the world size of zero is a subset of dp world size. For smaller models, it is usually a better choice to split the parameters within nodes with a setting ``parallel.zero1 <= 8``.
+
+Furthermore, you can enable communication-computation overlap, bucket reduce operation, gradient clipping in the config file.
+
+.. code-block:: python
+
+    hybrid_zero_optimizer = dict(
+        # Enable low_level_optimzer overlap_communication
+        zero_overlap_communication=True,
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:95]: 这里的参数配置有更新，可以看一下main分支最新的config配置

File: internlm/monitor/monitor.py [hunk 6]
```diff
@@ -211,6 +211,14 @@ def stop_monitor(self):
 
 @contextmanager
 def initialize_monitor_manager(job_name: str = None, alert_address: str = None):
+    """
+    Initialize monitor manager for future monitor and alert functions
```
Comment (huangting4201) [inline internlm/monitor/monitor.py:215]: Initialize monitor manager for future monitor and alert functions要不改成Initialize monitor manager for monitoring training lifetime and alerting exception info to Feishu.

File: internlm/train/training_internlm.py [hunk 7]
```diff
@@ -125,7 +128,14 @@ def get_train_data_loader(
     """
     Generate and return the training data loader.
 
-    Returns: A tuple of (train_dl, dataset_types).
+    Args:
+        num_worker (:class:`int`): number of subprocesses used for dataloader
```
Comment (huangting4201) [inline internlm/train/training_internlm.py:132]: 按照规范，参数注释结尾要加句号.

File: doc/code-docs/source/training.rst [hunk 8]
```diff
@@ -1,2 +1,10 @@
 Training API
-============
\ No newline at end of file
+============
+
+InternLM training API is managed in ``internlm.core.engine.Engine``. After defining the model, optimizer, criterion, train_dataloader, and lr_scheduler,
```
Comment (huangting4201) [inline doc/code-docs/source/training.rst:4]: 这里的Training API应该是针对internlm.core.trainer来写

File: doc/code-docs/source/initialize.rst [hunk 9]
```diff
@@ -20,16 +20,30 @@ parser with some builtin arguments, users can add custom parameters to this pars
 .. autofunction:: internlm.initialize.get_default_parser
 
 
-.. _InternLM-init:
+.. _InternLM-model-init:
 
 Model Initialization
 -------------------------
 
+.. autofunction:: internlm.train.initialize_model
```
Comment (huangting4201) [inline doc/code-docs/source/initialize.rst:28]: model初始化这里多介绍两点：
1. config里可以通过model_type字段指定要初始化的模型类型
2. 用户可使用utils/registry.py下的MODEL_INITIALIZER注册自定义的模型结构初始化函数，使用方式可参考：
```python
@MODEL_INITIALIZER.register_module(module_name=MODEL_TYPE)
def build_model_with_cfg(*args, **kwargs):
```

File: doc/code-docs/source/parallel.rst [hunk 10]
```diff
@@ -1,23 +1,123 @@
...
+        2. interleaved_overlap: bool, enable/disable communication overlap when using interleaved pipeline scheduler.
+    tensor parallel: tensor parallel size, usually the number of GPUs per node.
+    """
+    parallel = dict(
+        zero1=8,
+        pipeline=dict(size=1, interleaved_overlap=True),
+        sequence_parallel=False,
+    )
+
+Note: `Total number of GPUs = tensor parallel size * pipeline parallel size * data parallel size`
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:33]: 这块替换成这样吧：
Training parallel configuration example:

```python
parallel = dict(
    zero1=8,
    tensor=1,
    pipeline=dict(size=1, interleaved_overlap=True),
    sequence_parallel=False,
)
```

- zero1: zero parallel strategy, divided into the following three cases, default value is -1
  - When `size <= 0`, the size of the zero1 process group is equal to the size of the data parallel process group, so the optimizer state parameters will be split within the data parallel range.
  - When `size == 1`, zero1 is not used, and all data parallel groups retain the complete optimizer state parameters.
  - When `size > 1` and `size <= data_parallel_world_size`, the zero1 process group is a subset of the data parallel process group.
- tensor: tensor parallel size, usually the number of GPUs per node, default is 1
- pipeline: pipeline parallel strategy
   - size: pipeline parallel size, the default value is 1
   - interleaved_overlap: bool type, when interleaved scheduling, enable or disable communication optimization, the default value is False
- sequence_parallel: Whether to enable sequence parallelism, the default value is False

Note: `Data parallel size = Total number of GPUs / Pipeline parallel size / Tensor parallel size`

File: doc/code-docs/source/parallel.rst (hunk 1 above)
Comment (zigzagcai) [reply doc/code-docs/source/parallel.rst]: 已改为英文注释

File: internlm/train/training_internlm.py (hunk 7 above)
Comment (zigzagcai) [reply internlm/train/training_internlm.py]: 已修复

File: doc/code-docs/source/training.rst (hunk 8 above)
Comment (zigzagcai) [reply doc/code-docs/source/training.rst]: 已更新

File: doc/code-docs/source/initialize.rst (hunk 9 above)
Comment (zigzagcai) [reply doc/code-docs/source/initialize.rst:66]: 已更新model初始化部分，新增了这三点介绍

File: doc/code-docs/source/parallel.rst (hunk 10 above)
Comment (zigzagcai) [reply doc/code-docs/source/parallel.rst:36]: 已更新

File: doc/code-docs/source/parallel.rst (hunk 2 above)
Comment (zigzagcai) [reply doc/code-docs/source/parallel.rst:35]: 已把parallel配置贴出，并且增加说明

File: doc/code-docs/source/parallel.rst [hunk 11]
```diff
@@ -1,23 +1,130 @@
...
+        reduce_bucket_size=512 * 1024 * 1024,
+        # grad clipping
+        clip_grad_norm=1.0,
+    )
+
+.. autoclass:: internlm.solver.optimizer.hybrid_zero_optim.HybridZeroOptimizer
+    :members:
+
+.. [#f1] FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness
+.. [#f2] Efficient Large-Scale Language Model Training on GPU Clusters Using Megatron-LM
```
Comment (huangting4201) [inline doc/code-docs/source/parallel.rst:130]: 这个不是链接地址？

File: doc/code-docs/source/parallel.rst (hunk 5 above)
Comment (zigzagcai) [reply doc/code-docs/source/parallel.rst]: 已按照main分支最新配置更新docs

File: doc/code-docs/source/parallel.rst [hunk 12]
```diff
@@ -3,21 +3,101 @@ Parallel Training
...
+To use non-interleaved pipeline scheduler, you need to set ``model.num_chunks = 1`` in the config file.
+
+.. autoclass:: internlm.core.scheduler.pipeline_scheduler.PipelineScheduler
+    :members:
+
+scheduler for interleaved 1F1B strategy
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To use interleaved pipeline scheduler, you need to set ``model.num_chunks > 1`` in the config file.
+
+Also, to enable communication overlap when using interleaved pipeline scheduler, you need to set ``parallel.pipeline.interleaved_overlap = True``
```
[the remark this answers is not in the document]
Comment (zigzagcai) [reply doc/code-docs/source/parallel.rst:79]: 已添加关于interleaved_overlap的优化介绍

**Review by huangting4201 — APPROVED** [review]

**Pull request merged**